fl4health/model_bases/pca.py

```diff
@@ -0,0 +1,166 @@
+from logging import INFO
+from typing import Optional, Tuple
+
+import torch
+import torch.nn as nn
+from flwr.common.logger import log
+from torch import Tensor
+from torch.nn.parameter import Parameter
+
+
+class PCAModule(nn.Module):
+    def __init__(self, low_rank: bool = True, full_svd: bool = False, rank_estimation: int = 6) -> None:
+        """
+        PyTorch module for performing Principal Component Analysis.
+
+        Args:
+            low_rank (bool, optional): Indicates whether the data matrix is low-rank. If the user has
+             prior knowledge that it is, then this parameter can be set to True to allow for more efficient
+              computation of SVD. Defaults to True.
+            full_svd (bool, optional): Indicates whether full SVD or reduced SVD is performed. Defaults to False.
+            rank_estimation (int, optional): An estimation of the rank of the data matrix.
+            Only used if self.low_rank is True. Defaults to 6.
+        """
+        super().__init__()
+        self.low_rank = low_rank
+        self.full_svd = full_svd
+        self.rank_estimation = rank_estimation
+        self.principal_components: Parameter
+        self.singular_values: Parameter
+        self.data_mean: Tensor
+
+    def forward(self, X: Tensor, center_data: bool) -> Tuple[Tensor, Tensor]:
+        """
+        Perform PCA on the data matrix X by computing its SVD.
+
+        Args:
+            X (Tensor): Data matrix.
+            center_data (bool): If true, then the data mean will be subtracted from all data points prior to
+            performing PCA.
+
+        Returns:
+            Tuple[Tensor, Tensor]: The principal components (i.e., left singular vectors) and their corresponding
+            singular values.
+        """
+        X_prime = self.prepare_data_forward(X, center_data=center_data)
+        if self.low_rank:
+            log(INFO, "Assuming data matrix is low rank, using low-rank PCA implementation.")
+            q = min(self.rank_estimation, X_prime.size(0), X_prime.size(1))
+            _, S, V = torch.pca_lowrank(X_prime, q=q, center=False)
```

Comment:
I think you can just do
```
_, singular_values, principal_components = torch.pca_lowrank(X_prime, q=q, center=False)
```
Similarly below I think you can do
```
_, singular_values, Vh = torch.linalg.svd(X_prime, full_matrices=self.full_svd)
principal_components = Vh.T
```
and get ride of the `singular_values = S` on line 58